### PR TITLE
List Index expression

### PR DESCRIPTION
**Pull request by AlmanaX-21**
Big shoutout to GitHub copilot, stack overflow and some addons I checked out for code.

### CODE REVIEW

File: src/main/java/ch/njol/skript/expressions/base/ExprListIndex.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+			Object o = ((Variable<?>) listExpr).getRaw(e);
+			if (o instanceof TreeMap<?, ?>) {
+				TreeMap<?, ?> listVar = new TreeMap<>();
+				@SuppressWarnings("unchecked")
+				Stream<String> lazyIndexes = ((TreeMap<String, ?>) (first ? listVar : listVar.descendingMap())).entrySet().stream().filter(entry -> entry.getValue().equals(element)).map(Map.Entry::getKey);
+				return all ? lazyIndexes.toArray(String[]::new) : lazyIndexes.findFirst().map(CollectionUtils::array).orElse(null);
+			} else return null;
+		} else {
+			Object[] array = listExpr.getArray(e);
+			if (all) return LongStream.range(0, array.length).filter(i -> element.equals(array[(int) i])).boxed().toArray(Long[]::new);
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/base/ExprListIndex.java:49]: Skript coding standards puts singleton conditions on a new line

File: src/main/java/ch/njol/skript/expressions/base/ExprListIndex.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import java.util.TreeMap;
+import java.util.stream.LongStream;
+import java.util.stream.Stream;
+
+public class ExprListIndex extends SimpleExpression<Object> {
+
+	static {
+		Skript.registerExpression(ExprListIndex.class, Object.class, ExpressionType.COMBINED,
+			"[(first|2¦last)] [(1¦variable)] index of %object% in %objects%",
+			"[all] [(1¦variable)] ind(ex|ic)es of %object% in %objects%");
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/base/ExprListIndex.java:26]: We could add support for multiple objects in the second syntaxes index search

**Review by TheLimeGlass — CHANGES_REQUESTED** [review]

**Comment by TheLimeGlass** [comment]
I have enabled the workflow actions for you and future commits.

**Comment by TheLimeGlass** [comment]
GitHub actions is complaining because you are missing the license header. Just run `./gradlew licenseFormat` and Gradle will place it for you.

File: src/main/java/ch/njol/skript/expressions/base/ExprListIndex.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+		return variable ? String.class : Long.class;
+	}
+
+	@Override
+	public @NotNull String toString(@Nullable Event e, boolean debug) {
+		return (all ? "all" : first ? "first" : "last") + (variable ? " variable" : "") + (isSingle() ? " index" : " indexes") + " of " + elementExpr.toString(e, debug) + " in " + listExpr.toString(e, debug);
+	}
+
+	@Override
+	public boolean init(Expression<?> @NotNull [] exprs, int matchedPattern, @NotNull Kleenean isDelayed, @NotNull SkriptParser.ParseResult parseResult) {
```
[the remark this answers is not in the document]
Comment (AlmanaX-21) [reply src/main/java/ch/njol/skript/expressions/base/ExprListIndex.java:71]: 1. Done

[the remark this answers is not in the document]
Comment (AlmanaX-21) [reply src/main/java/ch/njol/skript/expressions/base/ExprListIndex.java:71]: Done

File: src/main/java/ch/njol/skript/expressions/base/ExprListIndex.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	}
+
+	private Expression<?> elementExpr, listExpr;
+
+	private boolean variable, all, first;
+
+
+	@Override
+	protected @Nullable
+	Object[] get(@NotNull Event e) {
```
[the remark this answers is not in the document]
Comment (AlmanaX-21) [reply src/main/java/ch/njol/skript/expressions/base/ExprListIndex.java:36]: done

File: src/main/java/ch/njol/skript/expressions/base/ExprListIndex.java (hunk 2 above)
Comment (AlmanaX-21) [reply src/main/java/ch/njol/skript/expressions/base/ExprListIndex.java:26]: I will leave you guys to it

File: build.gradle [hunk 5]
```diff
@@ -26,6 +26,7 @@ allprojects {
 }
 
 dependencies {
+	implementation 'org.jetbrains:annotations:20.1.0'
```
Comment (skmedix) [inline build.gradle:29]: Not needed, Skript uses `org.eclipse.jdt.annotation.Nullable`

File: src/main/java/ch/njol/skript/expressions/ExprListIndex.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+import ch.njol.skript.lang.ExpressionType;
+import ch.njol.skript.lang.SkriptParser;
+import ch.njol.skript.lang.Variable;
+import ch.njol.skript.lang.util.SimpleExpression;
+import ch.njol.skript.util.LiteralUtils;
+import ch.njol.util.Kleenean;
+import ch.njol.util.coll.CollectionUtils;
+import org.bukkit.event.Event;
+
+import javax.annotation.Nullable;
```
Comment (skmedix) [inline src/main/java/ch/njol/skript/expressions/ExprListIndex.java:35]:
```suggestion
import org.eclipse.jdt.annotation.Nullable;
```

**Review by skmedix — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprListIndex.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+			return false;
+		if (variable && !(listExpr instanceof Variable)) {
+			Skript.error(listExpr + " is not a list variable");
+			return false;
+		}
+		return LiteralUtils.canInitSafely(elementExpr, listExpr);
+	}
+
+	@Override
+	protected @Nullable
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprListIndex.java:82]: Nullable goes above the protected modifier and below the Override on it's own line

File: src/main/java/ch/njol/skript/expressions/ExprListIndex.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+	Object[] get(Event e) {
+		Object element = elementExpr.getSingle(e);
+		if (element == null) return null;
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprListIndex.java:85]: Ensure this is on two lines with Skript's coding standard.

**Review by TheLimeGlass — CHANGES_REQUESTED** [review]

**Pull request closed without merging**